File: CS/WindowsFormsApplication1/XtraReport1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using System.Collections.Generic;

namespace WindowsFormsApplication1
{
    public partial class XtraReport1 : DevExpress.XtraReports.UI.XtraReport
    {
        List<Data> source;

        public XtraReport1(List<Data> source, int detailPrintCount)
        {
            InitializeComponent();
            this.source = source;
            this.ReportPrintOptions.DetailCountOnEmptyDataSource = detailPrintCount;
        }

        int printCount = -1;

        private void Detail_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
        {
            printCount++;
            Data data = null;

            if (printCount < source.Count) {
                data = source[printCount];
            }

            xrTableCell1.Text = data == null ? string.Empty : data.Name;
            xrTableCell2.Text = data == null ? string.Empty : data.Category;
            xrTableCell3.Text = data == null ? string.Empty : string.Format("{0:c2}", data.Price);
        }

    }
}

File: VB/WindowsFormsApplication1/XtraReport1.vb
Imports System.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports System.Collections.Generic

Namespace WindowsFormsApplication1

    Public Partial Class XtraReport1
        Inherits XtraReport

        Private source As List(Of Data)

        Public Sub New(ByVal source As List(Of Data), ByVal detailPrintCount As Integer)
            InitializeComponent()
            Me.source = source
            ReportPrintOptions.DetailCountOnEmptyDataSource = detailPrintCount
        End Sub

        Private printCount As Integer = -1

        Private Sub Detail_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs)
            printCount += 1
            Dim data As Data = Nothing
            If printCount < source.Count Then
                data = source(printCount)
            End If

            xrTableCell1.Text = If(data Is Nothing, String.Empty, data.Name)
            xrTableCell2.Text = If(data Is Nothing, String.Empty, data.Category)
            xrTableCell3.Text = If(data Is Nothing, String.Empty, String.Format("{0:c2}", data.Price))
        End Sub
    End Class
End Namespace
